File: iteration2/src/MessagesInterface.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.HashMap;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import java.util.Scanner;

public class MessagesInterface {

    Colors Colors = new Colors();

    private Object messageObject;
    private JSONArray messageJson;
    private Session session;

    private ArrayList<Message> receivedMessages = new ArrayList<>();
    private ArrayList<Message> sentMessages = new ArrayList<>();
    private Scanner scanner;

    public MessagesInterface(Session session) {
        this.session = session;
        try {
            messageObject = new JSONParser().parse(new FileReader("./jsons/messages.json"));
            messageJson = (JSONArray) messageObject;
        } catch (Exception e) {
            System.out.println("Error: " + e);
        }
        scanner = new Scanner(System.in);

    }

    public void messagesMenu() {
        int choice = 0;
        while (true) {
            System.out.println(Colors.getRED() + Colors.getBOLD() + "\n>> Messages Menu\n" + Colors.getRESET());
            System.out
                    .println(Colors.getYELLOW() + "1" + Colors.getRESET() + ".   " + Colors.getGREEN() + "New Message"
                            + Colors.getRESET());
            System.out.println(Colors.getYELLOW() + "2" + Colors.getRESET() + ".   Received Messages");
            System.out.println(Colors.getYELLOW() + "3" + Colors.getRESET() + ".   Sent Messages");
            System.out
                    .println(Colors.getYELLOW() + "0" + Colors.getRESET()
                            + ".   Go back to the Course Registration System Menu");
            System.out.print("\n" + Colors.getBLUE() + "--> " + Colors.getRESET() + "What do you want to do?   ");
            System.out.print(Colors.getBLUE());
            choice = scanner.nextInt();
            System.out.print(Colors.getRESET());
            if (choice == 1) {
                System.out.print(Colors.getYELLOW() + "Enter receiver ID: " + Colors.getRESET());
                String receiver = scanner.next();
                System.out.println(Colors.getYELLOW() + "\nEnter subject " + Colors.getRESET()
                        + "(Press Enter twice to go to next line) : ");
                String subject = readMultiLineInput();
                System.out.println(Colors.getYELLOW() + "\nEnter description " + Colors.getRESET()
                        + "(Press Enter twice to go to next line) : ");
                String description = readMultiLineInput();
                newMessage(description, subject, receiver);
            } else if (choice == 2) {
                showReceivedMessages();
            } else if (choice == 3) {
                showSentMessages();
            } else if (choice == 0) {
                return;
            } else {
                System.out.println(Colors.getYELLOW() + "Invalid choice! Please try again." + Colors.getRESET());
            }

        }

    }

    private void calculateReceivedMessages() {

        for (Object messageObj : messageJson) {
            JSONObject message = (JSONObject) messageObj;
            String receiver = (String) message.get("Receiver");
            String sender = (String) message.get("Sender");

            if (receiver.equals(session.getUser().getID())) {
                String description = (String) message.get("Description");
                String subject = (String) message.get("Subject");

                Message receivedmessage = new Message(sender, receiver, description, subject);

                receivedMessages.add(receivedmessage);
            }
        }
    }

    private void showReceivedMessages() {
        calculateReceivedMessages();
        System.out.println(
                Colors.getRED()+Colors.getBOLD() + "\n>>> Received Messages\n" + Colors.getRESET()+Colors.getRESET());
        int i = 1;
        if (receivedMessages.size() == 0) {
            
            System.out.println(Colors.getYELLOW()+"You have no received messages.\n"+Colors.getRESET());
            return;
        }
        
        for (Message message : receivedMessages) { 
            System.out.println(Colors.getBLUE()+"Received Message " + i++ + ":" + Colors.getRESET());
            System.out.println(Colors.getYELLOW()+"\tSender: " +Colors.getRESET()+ message.getSenderID());
            System.out.println(Colors.getYELLOW()+"\tSubject: " +Colors.getRESET()+ message.getSubject());
            System.out.println(Colors.getYELLOW()+"\tDescription: " +Colors.getRESET()+ message.getDescription());
        }
        System.out.println("");
    }

    private void calculateSentMessages() {

        for (Object messageObj : messageJson) {
            JSONObject message = (JSONObject) messageObj;
            String receiver = (String) message.get("Receiver");
            String sender = (String) message.get("Sender");
            if (sender.equals(session.getUser().getID())) {
                String description = (String) message.get("Description");
                String subject = (String) message.get("Subject");

                Message sentMessage = new Message(sender, receiver, description, subject);

                sentMessages.add(sentMessage);
            }
        }

    }

    private void showSentMessages() {
        calculateSentMessages();
        int i=1;
        
        System.out
                .println(Colors.getRED()+ Colors.getBOLD()+ "\n>> Sent Messages\n" + Colors.getRESET()+Colors.getRESET());
        if (sentMessages.size() == 0) {
            System.out.println(Colors.getYELLOW()+"You have no sent messages.\n"+Colors.getRESET());
            return;
        }

        for (Message message : sentMessages) {
            System.out.println(Colors.getBLUE()+"Sent Message " + i++ + ":" + Colors.getRESET());
            System.out.println(Colors.getYELLOW()+"\tReceiver: " +Colors.getRESET()+ message.getReceiverID());
            System.out.println(Colors.getYELLOW()+"\tSubject: " +Colors.getRESET()+ message.getSubject());
            System.out.println(Colors.getYELLOW()+"\tDescription: " +Colors.getRESET()+ message.getDescription());
        }
        System.out.println("");
    }

    // send message specific person
    private void newMessage(String description, String subject, String receiver) {
        JSONObject newMessage = new JSONObject();
        newMessage.put("Sender", session.getUser().getID());
        newMessage.put("Receiver", receiver);
        newMessage.put("Subject", subject);
        newMessage.put("Description", description);

        messageJson.add(newMessage);

        try {
            PrintWriter pw = new PrintWriter("./jsons/messages.json");
            pw.write(messageJson.toJSONString());

            pw.flush();
            pw.close();

        } catch (Exception e) {
            System.out.println(Colors.getYELLOW() + "Error: " + e + Colors.getRESET());
        }
    }

    private String readMultiLineInput() {
        StringBuilder sb = new StringBuilder();
        String line;

        try {
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            while ((line = br.readLine()) != null && !line.trim().isEmpty()) {
                sb.append(line).append("\n");
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
        return sb.toString().trim();

    }
}

File: iteration3/src/messages_interface.py
import json
from typing import List

from colors import Colors
from session import Session
from message import Message

class MessagesInterface:
    def __init__(self, session: Session):
        self.__colors = Colors()
        self.__message_object = None
        self.__message_json = None
        self.__session = session
        self.__received_messages: List[Message] = []
        self.__sent_messages: List[Message] = []

        try:
            with open("./jsons/messages.json") as file:
                self.__message_json = json.load(file)
        except Exception as e:
            print(f"Error: {e}")

    def messages_menu(self):
        choice = 0
        while True:
            print(self.__colors.get_red() + self.__colors.get_bold() + "\n>> Messages Menu\n" + self.__colors.get_reset())
            print(self.__colors.get_yellow() + "1" + self.__colors.get_reset() + ".   " + self.__colors.get_green() +
                  "New Message" + self.__colors.get_reset())
            print(self.__colors.get_yellow() + "2" + self.__colors.get_reset() + ".   Received Messages")
            print(self.__colors.get_yellow() + "3" + self.__colors.get_reset() + ".   Sent Messages")
            print(self.__colors.get_yellow() + "0" + self.__colors.get_reset() +
                  ".   Go back to the Course Registration System Menu")

            print("\n" + self.__colors.get_blue() + "--> " + self.__colors.get_reset() +
                  "What do you want to do?   ", end="")
            choice = int(input().strip())

            if choice == 1:
                print(self.__colors.get_yellow() + "Enter receiver ID: " + self.__colors.get_reset(), end="")
                receiver = input().strip()
                print(self.__colors.get_yellow() + "\nEnter subject " + self.__colors.get_reset() +
                      "(Press Enter twice to go to the next line): ")
                subject = self.__read_multi_line_input()
                print(self.__colors.get_yellow() + "\nEnter description " + self.__colors.get_reset() +
                      "(Press Enter twice to go to the next line): ")
                description = self.__read_multi_line_input()
                self.__new_message(description, subject, receiver)
            elif choice == 2:
                self.__show_received_messages()
            elif choice == 3:
                self.__show_sent_messages()
            elif choice == 0:
                return
            else:
                print(self.__colors.get_yellow() + "Invalid choice! Please try again." + self.__colors.get_reset())

    def __calculate_received_messages(self):
        for message_obj in self.__message_json:
            message = json.loads(message_obj)
            receiver = message.get("Receiver")
            sender = message.get("Sender")

            if receiver == self.__session.get_user().get_id():
                description = message.get("Description")
                subject = message.get("Subject")

                received_message = Message(sender, receiver, description, subject)
                self.__received_messages.append(received_message)

    def __show_received_messages(self):
        self.__calculate_received_messages()
        print(self.__colors.get_red() + self.__colors.get_bold() +
              "\n>>> Received Messages\n" + self.__colors.get_reset())
        i = 1
        if not self.__received_messages:
            print(self.__colors.get_yellow() + "You have no received messages.\n" + self.__colors.get_reset())
            return

        for message in self.__received_messages:
            print(self.__colors.get_blue() + f"Received Message {i}:" + self.__colors.get_reset())
            print(self.__colors.get_yellow() + f"\tSender: {message.get_sender_id()}" + self.__colors.get_reset())
            print(self.__colors.get_yellow() + f"\tSubject: {message.get_subject()}" + self.__colors.get_reset())
            print(self.__colors.get_yellow() + f"\tDescription: {message.get_description()}" + self.__colors.get_reset())
            i += 1
        print("")

    def __calculate_sent_messages(self):
        for message_obj in self.__message_json:
            message = json.loads(message_obj)
            receiver = message.get("Receiver")
            sender = message.get("Sender")
            if sender == self.__session.get_user().get_id():
                description = message.get("Description")
                subject = message.get("Subject")

                sent_message = Message(sender, receiver, description, subject)
                self.__sent_messages.append(sent_message)

    def __show_sent_messages(self):
        self.__calculate_sent_messages()
        i = 1
        print(self.__colors.get_red() + self.__colors.get_bold() +
              "\n>> Sent Messages\n" + self.__colors.get_reset())
        if not self.__sent_messages:
            print(self.__colors.get_yellow() + "You have no sent messages.\n" + self.__colors.get_reset())
            return

        for message in self.__sent_messages:
            print(self.__colors.get_blue() + f"Sent Message {i}:" + self.__colors.get_reset())
            print(self.__colors.get_yellow() + f"\tReceiver: {message.get_receiver_id()}" + self.__colors.get_reset())
            print(self.__colors.get_yellow() + f"\tSubject: {message.get_subject()}" + self.__colors.get_reset())
            print(self.__colors.get_yellow() + f"\tDescription: {message.get_description()}" + self.__colors.get_reset())
            i += 1
        print("")

    # this method creates a new message and adds it to the messages.json file
    def __new_message(self, description, subject, receiver):
        new_message = {
            "Sender": self.__session.get_user().get_id(),
            "Receiver": receiver,
            "Subject": subject,
            "Description": description,
        }

        self.__message_json.append(json.dumps(new_message))

        try:
            with open("./jsons/messages.json", "w") as file:
                json.dump(self.__message_json, file)
        except IOError as e:
            print(f"Error: {e}")

    def __read_multi_line_input(self):
        lines = []
        while True:
            line = input()
            if not line.strip():
                break
            lines.append(line)
        return "\n".join(lines)
